core/src/test/scala/integration/kafka/api/LogDirFailureTest.scala

```diff
@@ -64,6 +65,39 @@ class LogDirFailureTest extends IntegrationTestHarness {
   }
 
   @Test
+  def brokerWithOldIBPShouldHaltOnLogDirFailure() {
+    @volatile var statusCodeOption: Option[Int] = None
+    Exit.setHaltProcedure { (statusCode, _) =>
+      statusCodeOption = Some(statusCode)
+      throw new IllegalArgumentException
+    }
+
+    var server: KafkaServer = null
+    try {
+      val props = TestUtils.createBrokerConfig(serverCount, zkConnect, logDirCount = 3)
+      props.put(KafkaConfig.InterBrokerProtocolVersionProp, "0.11.0")
+      props.put(KafkaConfig.LogMessageFormatVersionProp, "0.11.0")
+      val kafkaConfig = KafkaConfig.fromProps(props)
+      val logDir = new File(kafkaConfig.logDirs.head)
+      // Make log directory of the partition on the leader broker inaccessible by replacing it with a file
+      CoreUtils.swallow(Utils.delete(logDir))
+      logDir.createNewFile()
+      assertTrue(logDir.isFile)
+
+      server = TestUtils.createServer(kafkaConfig)
+      TestUtils.waitUntilTrue(() => {
+        server.logDirFailureChannel.offlineLogDirQueue.size() == 0
+      }, "timed out waiting for LogDirFailureHandler to process offline dir")
```

Comment:
When the queue size is 0, the event may not be fully processed yet. A more reliable way is probably to do the waitUntilTrue on checking statusCodeOption directly. Then, we don't need to expose logDirFailureChannel.offlineLogDirQueue.

Reply:
Good point. I only realized this after I was away from keyboard. I have updated the patch as suggested.